docs/proposals-accepted/202206-active-series-limiting-hashring.md

```diff
@@ -0,0 +1,226 @@
+
+## How
+
+Thanos Receive uses Prometheus TSDB and creates a separate TSDB database instance internally for each of its tenants. When a Receive replica gets a remote write request, it loops through the timeseries, hashes labels with tenant name as prefix and forwards remote write request to other Receive nodes. Upon receiving samples in a remote write request from a tenant, the Receive node appends the samples to the in-memory HEAD block of a tenant.
+
+We can leverage this fact, and generate statistics from the HEAD block, which can give us an accurate idea of the active or HEAD series of a tenant. This can also be exposed as a metric.
+
+Thus, any remote write request can be failed completely, with a 429 status code and appropriate error message, if it increases the number of active series above the configured limit for a tenant. Partially accepting write requests might lead to confusing results and error semantics, so we propose to avoid this and focus on retries from client-side.
+
+There are however a few challenges to this, as tenant metric data is distributed and replicated across multiple Thanos Receive replicas. Also, with a multi-replica setup, we have the concept of per-replica-tenant and per-tenant limits that can be defined as,
+
+* **Per-replica-tenant limit**: The limit imposed for active series per tenant, per replica of Thanos Receive. An initial implementation of this is already WIP in this [PR](https://github.com/thanos-io/thanos/pull/5333). This can also be treated as the active series limit for non-hashring topology or single replica Receive setups.
+
+* **Per-tenant limit**: The overall limit imposed for active series per tenant across all replicas of Thanos Receive. This is essentially what this proposal is for.
+
+## Proposal
+
+In general, we would need three measures to impose a limit,
+
+* The current count of active series for a tenant (across all replicas if it is a *per-tenant* limit)
+* The user configured limit (*per-tenant* and *per-replica-tenant* can be different). We can assume this would be available as a user flag and would be same for all tenants (initially)
+* The increase in the number of active series, when a new tenant [remote write request](https://github.com/prometheus/prometheus/blob/v2.36.1/prompb/remote.proto#L22) would be ingested (this can be optional as seen in [meta-monitoring approach](#meta-monitoring-based-validator)).
+
+There are a few ways in which we can achieve the outlined goals of this proposal and get the above measurements to impose a limit. The order of approaches is based on preference.
+
+### Meta-monitoring-based Receive Router Validation
+
+We could leverage a meta-monitoring solution which scrapes metrics from all Receivers as and then programmatically query for metrics like `prometheus_tsdb_head_series` and sum across tenants and instances periodically and limit based on that value (`latestCurrentSeries`).
```

Comment:
I am confused about what is the "meta-monitoring solution". Is it a single Receive, called "Validator", that will be querying all the other Receive instances to aggregates tenants information? 

What happens in hashring architectures of multiple Router/Ingestor instances where a request could arrive at any instance? Is each Router/Ingestor also a Validator? Or one of them is picked to be a validator?

Reply:
Meta-monitoring is an external system that monitors Thanos components that you are running. Monitoring for your monitoring. 🙂 

In the context of this proposal, this would mean any solution which is capable of scraping metrics from all your Receive instances. That could be a single Prometheus that scrapes metrics from the cluster you are running Thanos in, or something more elaborate. 

This would usually vary from use case to use case, which is why we only mention "meta-monitoring" instead of accurately describing some set up.

Reply:
> This would usually vary from use case to use case, which is why we only mention "meta-monitoring" instead of accurately describing some set up.

We don't need an accurate description, but a simple example. Like what you said below: 

> In the context of this proposal, this would mean any solution which is capable of scraping metrics from all your Receive instances. That could be a single Prometheus that scrapes metrics from the cluster you are running Thanos in, or something more elaborate.

This is so well written. IMHO, it should be part of the proposal so that every reader has the same understanding of what this "meta monitoring" system is.

Reply:
Will add!